src/geo_traits_impl/geometry_collection.rs

```diff
@@ -0,0 +1,116 @@
+use super::GeometryCollectionType;
+use geo_traits::{Dimensions, GeometryTrait};
+
+impl geo_traits::GeometryCollectionTrait for GeometryCollectionType {
+    type T = f64;
+    type GeometryType<'b>
+        = &'b crate::Geometry
+    where
+        Self: 'b;
+
+    fn dim(&self) -> Dimensions {
+        self.geometry(0).unwrap().dim()
+    }
+
+    fn geometries(
+        &self,
+    ) -> impl DoubleEndedIterator + ExactSizeIterator<Item = Self::GeometryType<'_>> {
+        self.0.iter()
+    }
+
+    fn geometry(&self, i: usize) -> Option<Self::GeometryType<'_>> {
+        self.0.get(i)
+    }
+
+    unsafe fn geometry_unchecked(&self, i: usize) -> Self::GeometryType<'_> {
```

Comment:
This doesn't seem to agree with my understanding of our discussion https://github.com/georust/geo/pull/1157#discussion_r1815796902 

@kylebarron was using this method as *safe but could panic*. It's not marked unsafe in the trait.

@frewsxcv - I liked your suggestion of having both `foo_or_panic` and `foo _unchecked` flavors.

We should probably address this to avoid confusion in the future.

(note this comment applies to all the `unsafe` methods in this PR)

Reply:
Oh, apparently I missed the memo. `foo_unchecked` __is__ marked unsafe in geo-traits.

Reply:
No wait! It gets weirder:

```
// safe, but may panic
CoordTrait::nth_unchecked(&self, n: usize) -> Self::T;

// unsafe
GeometryCollectionTrait::geometry_unchecked
MultiPoint::point_unchecked
LineString::coord_unchecked
MultiLineString::line_string_unchecked
Polygon::interior_unchecked
MultiPolygon::polygon_unchecked
```

So it seems like CoordTrait is the outlier here.

Reply:
Yeah we can definitely change this in `geo_traits` so they are more aligned

Reply:
Yeah, sorry I really don't know the right semantics around these. I guess there's three variants?

1. `unchecked` will be UB if out of bounds?
2. `panic` (like a `Vec::get().unwrap()`?) that will do a bounds check and panic
3. `checked` that returns an `Option<T>`

Happy to make any changes.

Reply:
>  I guess there's three variants

That sounds right to me. I've just now opened a PR to rename the `CoordTrait::nth_unchecked` to `CoordTrait::nth_or_panic` to match this scheme.  https://github.com/georust/geo/pull/1242